meta-balena-common/classes/kernel-balena.bbclass

```diff
@@ -221,6 +221,16 @@ BALENA_CONFIGS[firmware_compress] = " \
     CONFIG_FW_LOADER_COMPRESS=y \
 "
 
+MODULE_COMPRESS = "${@configure_from_version("5.13", "module_compress", "", d)}"
+BALENA_CONFIGS[module_compress] = " \
+    CONFIG_MODULE_COMPRESS_ZSTD=y \
+"
+
+MODULE_DECOMPRESS = "${@configure_from_version("5.17", "module_decompress", "", d)}"
+BALENA_CONFIGS[module_decompress] = " \
+    CONFIG_MODULE_DECOMPRESS=y \
```

Comment:
@jakogut does this mean that older user space with kmod that do not support compressed modules but on kernel versions >=5.17 will still work?

Reply:
@alexgg I suspected this was the case, but I also wondered if the kmod tools needed any changes to recognize that a compressed kernel module could be loaded without first decompressing, so I tested it. This is a debian bullseye container that ships kmod without zstd support, as shown by the version string.
```
root@aff557b665f2:~# insmod -V
kmod version 28
-ZSTD +XZ -ZLIB +LIBCRYPTO -EXPERIMENTAL
root@aff557b665f2:~# insmod test.ko.zst 
insmod: ERROR: could not insert module test.ko.zst: Invalid module format
```

So it seems that at least older kmod versions don't recognize the compressed module can be loaded directly. I built kmod v32 on the same system to see if that would work, and it still doesn't load a compressed module.

Looking at the [Kconfig description](https://cateee.net/lkddb/web-lkddb/MODULE_DECOMPRESS.html) for `MODULE_DECOMPRESS`, it mentions that the feature is useful when the [load pinning security policy](https://docs.kernel.org/admin-guide/LSM/LoadPin.html) is enabled. This is apparently a feature that ensures that all modules and firmware are loaded from the same filesystem, without needing to sign individual files.

It seems kmod could be patched at this point to support loading compressed module files without first decompressing them, but it's not supported currently.

Reply:
I tracked down the original [patch](https://patchwork.kernel.org/project/linux-modules/patch/YaMYJv539OEBz5B/@google.com/), and the author describes the intent to keep the load pinning policy happy, as well as including a patch for kmod to make it support loading a compressed module.

Unfortunately, this puts us right back where we were. Kmod needs a patch to support loading compressed kernel modules, and old builds won't work with new modules.

Reply:
It looks like the patch has actually be [upstreamed](https://github.com/kmod-project/kmod/commit/30077bf1719eb9625b1c223c144b74285425d581) as of kmod v31, but it still seems to not work as expected.

Reply:
It seems kmod's support for passing off a compressed module to the kernel to decompress is disabled if userspace decompression is disabled, which seems a bit pointless. This looks worth patching upstream.

Reply:
Seems that there's no point in adding this setting at this point. We won't enable load pinning as that would prevent to load modules and firmware from containers, and it makes no difference with regards to user space.
I wouldn't worry about older user space, just remember that this change needs to be a major change as it could potentially break applications on older/discontinued operating systems that do not support kmod module compression.

Reply:
Agreed. I've marked this change as a major and dropped kernel decompression.

It seems in-kernel decompression is used by a new enough kmod version when compiled with the required options. Briefly looking at the source, I can't see why it doesn't use kernel decompression when the tool is compiled without userspace decompression.

It could still be handy in the future for seamlessly supporting new compression codecs without userspace support.